modules/tensor_mechanics/src/materials/ComputeThermalExpansionEigenstrainBase.C

```diff
@@ -10,41 +10,62 @@
 #include "RankTwoTensor.h"
 
+template <bool is_ad>
 InputParameters
-ComputeThermalExpansionEigenstrainBase::validParams()
+ComputeThermalExpansionEigenstrainBaseTempl<is_ad>::validParams()
 {
-  InputParameters params = ComputeEigenstrainBase::validParams();
+  InputParameters params = ComputeEigenstrainBaseTempl<is_ad>::validParams();
   params.addCoupledVar("temperature", "Coupled temperature");
   params.addRequiredCoupledVar("stress_free_temperature",
                                "Reference temperature at which there is no "
                                "thermal expansion for thermal eigenstrain "
                                "calculation");
+  params.addParam<bool>("use_old_temperature",
+                        false,
+                        "Flag to optionally use the temperature value from the previous timestep.");
   return params;
 }
 
-ComputeThermalExpansionEigenstrainBase::ComputeThermalExpansionEigenstrainBase(
+template <bool is_ad>
+ComputeThermalExpansionEigenstrainBaseTempl<is_ad>::ComputeThermalExpansionEigenstrainBaseTempl(
     const InputParameters & parameters)
-  : DerivativeMaterialInterface<ComputeEigenstrainBase>(parameters),
-    _temperature(coupledValue("temperature")),
-    _deigenstrain_dT(declarePropertyDerivative<RankTwoTensor>(_eigenstrain_name,
-                                                              getVar("temperature", 0)->name())),
-    _stress_free_temperature(coupledValue("stress_free_temperature"))
+  : DerivativeMaterialInterface<ComputeEigenstrainBaseTempl<is_ad>>(parameters),
+    _use_old_temperature(this->template getParam<bool>("use_old_temperature")),
+    _temperature(this->template coupledGenericValue<is_ad>("temperature")),
+    _temperature_old(this->_fe_problem.isTransient() ? this->coupledValueOld("temperature")
+                                                     : this->_zero),
+    _deigenstrain_dT(is_ad ? nullptr
+                           : &this->template declarePropertyDerivative<RankTwoTensor>(
+                                 _eigenstrain_name, this->getVar("temperature", 0)->name())),
+    _stress_free_temperature(this->coupledValue("stress_free_temperature"))
 {
+  if (_use_old_temperature && !this->_fe_problem.isTransient())
+    this->paramError(
+        "use_old_temperature",
+        "The old state of the temperature variable is only available in a transient simulation.");
 }
 
+template <bool is_ad>
 void
-ComputeThermalExpansionEigenstrainBase::computeQpEigenstrain()
+ComputeThermalExpansionEigenstrainBaseTempl<is_ad>::computeQpEigenstrain()
 {
-  Real thermal_strain = 0.0;
+  GenericReal<is_ad> thermal_strain = 0.0;
 
-  // instantaneous_cte is just the derivative of thermal_strain with respect to temperature
-  Real instantaneous_cte = 0.0;
+  if constexpr (is_ad)
+    computeThermalStrain(thermal_strain);
+  else
+  {
+    // instantaneous_cte is just the derivative of thermal_strain with respect to temperature
```

Comment:
Unrelated to this PR, it'd appear this should be clear at this point (but it isn't)

Reply:
I'd be in favor of renaming this parameter to dthermal_strain_dT everywhere

Reply:
The other name is too cryptic to me.